feat(gemini): add complete Gemini grounding tools support

Silentely opened the pull request:
## Summary

This PR adds complete support for Gemini's grounding tools in the AxonHub transformer, enabling seamless integration with Google's Gemini API grounding capabilities.

### New Features

- **googleSearch**: Web search grounding tool
- **codeExecution**: Code execution grounding tool  
- **urlContext**: URL content fetching grounding tool

### Changes

#### Files Modified (7 files, +105/-2 lines)

| File | Description |
|------|-------------|
| `internal/llm/constants.go` | Add `ToolTypeUrlContext` constant |
| `internal/llm/tools.go` | Add `UrlContext` type and field to unified Tool struct |
| `internal/llm/transformer/gemini/model.go` | Add `UrlContext` field to Gemini Tool model |
| `internal/llm/transformer/gemini/inbound_convert.go` | Add Gemini → LLM format conversion for urlContext |
| `internal/llm/transformer/gemini/outbound_convert.go` | Add LLM → Gemini format conversion for urlContext |
| `internal/llm/transformer/gemini/inbound_convert_test.go` | Add unit tests for inbound conversion |
| `internal/llm/transformer/gemini/outbound_convert_test.go` | Add unit tests for outbound conversion |

### Testing

#### Unit Tests
```
=== RUN   TestConvertGeminiToLLMRequest_Tools/request_with_url_context_tool
--- PASS: TestConvertGeminiToLLMRequest_Tools/request_with_url_context_tool (0.00s)
=== RUN   TestConvertGeminiToLLMRequest_Tools/request_with_all_grounding_tools
--- PASS: TestConvertGeminiToLLMRequest_Tools/request_with_all_grounding_tools (0.00s)
=== RUN   TestConvertLLMToGeminiRequest_Tools/request_with_url_context_tool
--- PASS: TestConvertLLMToGeminiRequest_Tools/request_with_url_context_tool (0.00s)
```

#### Integration Tests

**urlContext Tool** - Successfully fetched and summarized web content:
```json
{
  "contents": [{"role": "user", "parts": [{"text": "请简要总结这个网页的内容：https://ai.google.dev/gemini-api/docs/url-context"}]}],
  "tools": [{"urlContext": {}}]
}
// Response: Successfully returned summary of the URL content
```

**googleSearch Tool** - Successfully searched and returned news:
```json
{
  "contents": [{"role": "user", "parts": [{"text": "请搜索今天的科技新闻"}]}],
  "tools": [{"googleSearch": {}}]
}
// Response: Successfully returned recent tech news
```

### API Usage Example

```bash
# Using urlContext
curl -X POST "http://localhost:8090/gemini/v1beta/models/gemini-2.5-flash:generateContent" \
  -H "Authorization: Bearer YOUR_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "contents": [{"role": "user", "parts": [{"text": "Summarize: https://example.com"}]}],
    "tools": [{"urlContext": {}}]
  }'

# Using googleSearch
curl -X POST "http://localhost:8090/gemini/v1beta/models/gemini-2.5-flash:generateContent" \
  -H "Authorization: Bearer YOUR_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "contents": [{"role": "user", "parts": [{"text": "Search for latest AI news"}]}],
    "tools": [{"googleSearch": {}}]
  }'
```

```
{
  "contents": [
    {
      "parts": [
        {
          "text": "请搜索今天的科技新闻，告诉我最近有什么重大科技事件"
        }
      ],
      "role": "user"
    }
  ],
  "tools": [
    {
      "googleSearch": {}
    }
  ]
}
{
  "candidates": [
    {
      "content": {
        "parts": [
          {
            "text": "以下是近期的一些重大科技事件和新闻：\n\n**人工智能领域进展：**\n\n*   《时代》杂志将“人工智能的架构师”评为2025年度人物，凸显了人工智能在科技领域的中心地位。\n*   谷歌翻译通过整合Gemini技术，增强了实时翻译功能，并为耳机带来了新的实时语音到语音翻译功能。\n*   亚马逊推出了“前沿代理”、新型芯片以及AWS re:Invent大会上的私有“AI工厂”等产品。\n*   迪士尼向OpenAI投资10亿美元，旨在将米老鼠等迪士尼角色引入Sora AI视频工具。\n*   美国前总统特朗普签署了一项行政命令，旨在限制各州对人工智能的监管。\n*   亚马逊Kindle的新应用内AI功能允许用户就书籍内容提问。\n*   有观点认为，虽然许多工作面临被AI颠覆的风险，但AI培训师、AI审计师和AI集成商等角色仍将保持高度相关性。\n*   OpenAI正在调整新员工的股权“归属期”条款，以留住顶尖的AI人才。\n\n**其他科技新闻：**\n\n*   上诉法院支持对苹果公司的藐视法庭裁决，但重新开启了iPhone应用收费的讨论。\n*   加密货币大亨Do Kwon因400亿美元的稳定币欺诈案被判处15年监禁。\n*   福特公司正加大技术投入并与警方合作，以阻止F-150皮卡盗窃案。\n*   Reddit对澳大利亚禁止16岁以下儿童使用社交媒体的全球首例法律提出挑战。\n*   OnePlus 15R手机预计将于12月17日发布。\n*   Oppo A6x已在印度发布，配备6,500毫安时电池，起售价为12,499卢比。\n*   iPhone 16的价格在Croma上降价超过17,000卢比。\n\n**近期和即将举行的科技活动：**\n\n*   **QCon AI** 将于12月16-17日在纽约举行。\n*   **混沌通信大会 (Chaos Communication Congress)** 将于12月27-30日在汉堡举行。\n*   **CES 2026** 是全球最具影响力的科技盛会，将于2026年1月6日至9日在拉斯维加斯举行，展示突破性技术和全球创新。\n*   **GITEX GLOBAL 2025** 是全球最大的科技与人工智能盛会，第45届将于2025年10月13日在迪拜世界贸易中心和迪拜港举行。\n*   **新加坡科技周 (Tech Week Singapore)** 是亚洲最重要的科技盛会，将于2025年10月8日至9日在滨海湾金沙举行。\n*   **Web Summit** 将于2026年11月9日至12日返回里斯本。\n*   **LEAP 2026** 科技大会将于2026年4月13日至16日举行，同期还将举办DeepFest，这是该地区最具影响力的AI活动。"
          }
        ],
        "role": "model"
      },
      "finishReason": "STOP",
      "index": 0,
      "groundingMetadata": {
        "searchEntryPoint": {
          "renderedContent": "<style>\n.container {\n  align-items: center;\n  border-radius: 8px;\n  display: flex;\n  font-family: Google Sans, Roboto, sans-serif;\n  font-size: 14px;\n  line-height: 20px;\n  padding: 8px 12px;\n}\n.chip {\n  display: inline-block;\n  border: solid 1px;\n  border-radius: 16px;\n  min-width: 14px;\n  padding: 5px 16px;\n  text-align: center;\n  user-select: none;\n  margin: 0 8px;\n  -webkit-tap-highlight-color: transparent;\n}\n.carousel {\n  overflow: auto;\n  scrollbar-width: none;\n  white-space: nowrap;\n  margin-right: -12px;\n}\n.headline {\n  display: flex;\n  margin-right: 4px;\n}\n.gradient-container {\n  position: relative;\n}\n.gradient {\n  position: absolute;\n  transform: translate(3px, -9px);\n  height: 36px;\n  width: 9px;\n}\n@media (prefers-color-scheme: light) {\n  .container {\n    background-color: #fafafa;\n    box-shadow: 0 0 0 1px #0000000f;\n  }\n  .headline-label {\n    color: #1f1f1f;\n  }\n  .chip {\n    background-color: #ffffff;\n    border-color: #d2d2d2;\n    color: #5e5e5e;\n    text-decoration: none;\n  }\n  .chip:hover {\n    background-color: #f2f2f2;\n  }\n  .chip:focus {\n    background-color: #f2f2f2;\n  }\n  .chip:active {\n    background-color: #d8d8d8;\n    border-color: #b6b6b6;\n  }\n  .logo-dark {\n    display: none;\n  }\n  .gradient {\n    background: linear-gradient(90deg, #fafafa 15%, #fafafa00 100%);\n  }\n}\n@media (prefers-color-scheme: dark) {\n  .container {\n    background-color: #1f1f1f;\n    box-shadow: 0 0 0 1px #ffffff26;\n  }\n  .headline-label {\n    color: #fff;\n  }\n  .chip {\n    background-color: #2c2c2c;\n    border-color: #3c4043;\n    color: #fff;\n    text-decoration: none;\n  }\n  .chip:hover {\n    background-color: #353536;\n  }\n  .chip:focus {\n    background-color: #353536;\n  }\n  .chip:active {\n    background-color: #464849;\n    border-color: #53575b;\n  }\n  .logo-light {\n    display: none;\n  }\n  .gradient {\n    background: linear-gradient(90deg, #1f1f1f 15%, #1f1f1f00 100%);\n  }\n}\n</style>\n<div class=\"container\">\n  <div class=\"headline\">\n    <svg class=\"logo-light\" width=\"18\" height=\"18\" viewBox=\"9 9 35 35\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\">\n      <path fill-rule=\"evenodd\" clip-rule=\"evenodd\" d=\"M42.8622 27.0064C42.8622 25.7839 42.7525 24.6084 42.5487 23.4799H26.3109V30.1568H35.5897C35.1821 32.3041 33.9596 34.1222 32.1258 35.3448V39.6864H37.7213C40.9814 36.677 42.8622 32.2571 42.8622 27.0064V27.0064Z\" fill=\"#4285F4\"/>\n      <path fill-rule=\"evenodd\" clip-rule=\"evenodd\" d=\"M26.3109 43.8555C30.9659 43.8555 34.8687 42.3195 37.7213 39.6863L32.1258 35.3447C30.5898 36.3792 28.6306 37.0061 26.3109 37.0061C21.8282 37.0061 18.0195 33.9811 16.6559 29.906H10.9194V34.3573C13.7563 39.9841 19.5712 43.8555 26.3109 43.8555V43.8555Z\" fill=\"#34A853\"/>\n      <path fill-rule=\"evenodd\" clip-rule=\"evenodd\" d=\"M16.6559 29.8904C16.3111 28.8559 16.1074 27.7588 16.1074 26.6146C16.1074 25.4704 16.3111 24.3733 16.6559 23.3388V18.8875H10.9194C9.74388 21.2072 9.06992 23.8247 9.06992 26.6146C9.06992 29.4045 9.74388 32.022 10.9194 34.3417L15.3864 30.8621L16.6559 29.8904V29.8904Z\" fill=\"#FBBC05\"/>\n      <path fill-rule=\"evenodd\" clip-rule=\"evenodd\" d=\"M26.3109 16.2386C28.85 16.2386 31.107 17.1164 32.9095 18.8091L37.8466 13.8719C34.853 11.082 30.9659 9.3736 26.3109 9.3736C19.5712 9.3736 13.7563 13.245 10.9194 18.8875L16.6559 23.3388C18.0195 19.2636 21.8282 16.2386 26.3109 16.2386V16.2386Z\" fill=\"#EA4335\"/>\n    </svg>\n    <svg class=\"logo-dark\" width=\"18\" height=\"18\" viewBox=\"0 0 48 48\" xmlns=\"http://www.w3.org/2000/svg\">\n      <circle cx=\"24\" cy=\"23\" fill=\"#FFF\" r=\"22\"/>\n      <path d=\"M33.76 34.26c2.75-2.56 4.49-6.37 4.49-11.26 0-.89-.08-1.84-.29-3H24.01v5.99h8.03c-.4 2.02-1.5 3.56-3.07 4.56v.75l3.91 2.97h.88z\" fill=\"#4285F4\"/>\n      <path d=\"M15.58 25.77A8.845 8.845 0 0 0 24 31.86c1.92 0 3.62-.46 4.97-1.31l4.79 3.71C31.14 36.7 27.65 38 24 38c-5.93 0-11.01-3.4-13.45-8.36l.17-1.01 4.06-2.85h.8z\" fill=\"#34A853\"/>\n      <path d=\"M15.59 20.21a8.864 8.864 0 0 0 0 5.58l-5.03 3.86c-.98-2-1.53-4.25-1.53-6.64 0-2.39.55-4.64 1.53-6.64l1-.22 3.81 2.98.22 1.08z\" fill=\"#FBBC05\"/>\n      <path d=\"M24 14.14c2.11 0 4.02.75 5.52 1.98l4.36-4.36C31.22 9.43 27.81 8 24 8c-5.93 0-11.01 3.4-13.45 8.36l5.03 3.85A8.86 8.86 0 0 1 24 14.14z\" fill=\"#EA4335\"/>\n    </svg>\n    <div class=\"gradient-container\"><div class=\"gradient\"></div></div>\n  </div>\n  <div class=\"carousel\">\n    <a class=\"chip\" href=\"https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQGzeDjBPDZXcJGD0ilSsnYiBO6aXhXMywse-e_mtcLJDlaDjEXr4u_VmdrI7bnLr9J7-LPsTcRdJlIgEfoK6kLQSiUpP-08q2T9fyxxoRIjTd_TEXcdaxf70IpmP1DmQ0A0Ys1Anuzi_bfNRnbZINLJn5cJfQavERjTKfMiSXdfK6IrbypTbCnaQtzfM5MZkNPgI2KiBUqc46VsRg==\">today&#39;s technology news</a>\n    <a class=\"chip\" href=\"https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQEGy1k6b4Qc9IpA14nTx8arrb_aTWikreq9XAfdlD1KUSCjk3MCy0329ovMwZ48jF2vHPiEhOt4jVjIINA5g-aTuh2SE9kR9zh70ixvRYyzIZswD1MAbHHybeYY2ET4wcWuCNYI_NaeZ4VtGE881AH5NERFQ8LhbfeC-OYy7pIuS9Y5Yuz8yHOxzgS4tShyNU_qM7GzawwEcviMlriXQ-8n4Yw=\">major technology events recently</a>\n  </div>\n</div>\n"
        },
        "groundingChunks": [
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQHZXFmwg-Xyh0hCQjkKx7Mn4WyNd2YvtWWRHFErR83mILKG7q8NAQVyC7sAMzo65aKvMkJnKw6aBQPmDokXU15Ts2qM-hl9vqQY4VfnOmLIZ5161zo3bhI=",
              "title": "apnews.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQHJhXVv1_l_QkykRtbSQuNEHUiwXTR-szpMU9W-3fOpTcIZuuJrtRmZNfgwkzzESgy7b_h7jZ2jyhnDZKVQEqOdrivtDKYkwfr8gqprvDLUQMTRUw7OGtYQ7sdsfTdR9D7bwyfeUG4t",
              "title": "thehindu.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQGNIqi7XsHJmaRFN13VPGH2wAV89Sakiw5RhIRjoq3P-NCOPZBrHwwHT0r6ZipHvCkvAQjJQvpU77hQgd_6NtADLm_FcwaYC2eQn1Va8TUQ8cIdRQ==",
              "title": "engadget.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQEB8gM1Fz5CddHLOQZcHI5FsAy5_1E4Y7an0hKQ8BNfgGIWiC_Y8Z_iKtU4VrIYzQx30J5s9iCaEXeqkP2KyV-dd3AQO6tVMqPgW7e_W57K3L6vzw==",
              "title": "geekwire.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQFQBoQAykj3UCNlcZ-v_bqHUangVuv9_dZynR3HWPJpZoaDF2ghCPMLF7jx_IXYTaa4fk6h0E2E6n5OqR36Dg2UpYE3-o2GgKxtrOEtBliKyKFquZ5OEM20R98vJLxh6f9K1h_oK05I",
              "title": "indianexpress.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQGtVITLlOD-wuJRWHXYFni0vqmv0io36b-VhZpYqsuuOdtUZh0JWQ_oDQ7yddw19Joj0ojtKqnq6z3937jtfAdmM-OCcmf9z5uLE96KV5N0ojjzIbWTL-Wpgw==",
              "title": "techmeme.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQElGvGSJjmH7120hatJ8Wgc3qR4wr6xFNYxHIrtuSnNQJA3J65TEXUXXNvz6RWq1qt1quwm0qKwrcX1r3imRKHlpdkjjazVPaKdPYHxAZfv",
              "title": "ces.tech"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQFWGmbJ15b5dXbIuQbp8ZLq1wz4_39vkYehtgiPeXtvGz3nFFqNawoi_tZQjNc_A6LvP52N61Ji-Ig8RniZrp8DtDwZRn-1yz9EXPfpQt4NT0aIDOsRihXpSQBtXy-ygg==",
              "title": "technologymagazine.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQEtWQlxmw_vBBl3hhJLLcvgphXDTg5VlKWMb6rO98Zju8YXdQZzm8g2Q3cLtdNqVyuVye5tIuX6738tUQ1sgxqwT3BdDFmox5NnWd9s0eFG-g==",
              "title": "websummit.com"
            }
          },
          {
            "web": {
              "uri": "https://vertexaisearch.cloud.google.com/grounding-api-redirect/AUZIYQEB21RzlsoSye7ZXjlfhYSrYMsD3VrZLhgX7qtnn1uUXqKp-rPwEwvbqm0EbHVk3Vn6zfxfy51NxK1XemgVKKEVWmdJ1Zgacb0rQ-YqV_p4jcoVcg==",
              "title": "onegiantleap.com"
            }
          }
        ],
        "groundingSupports": [
          {
            "segment": {
              "startIndex": 89,
              "endIndex": 223,
              "text": "*   《时代》杂志将“人工智能的架构师”评为2025年度人物，凸显了人工智能在科技领域的中心地位。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 224,
              "endIndex": 360,
              "text": "*   谷歌翻译通过整合Gemini技术，增强了实时翻译功能，并为耳机带来了新的实时语音到语音翻译功能。"
            },
            "groundingChunkIndices": [
              1,
              2
            ]
          },
          {
            "segment": {
              "startIndex": 361,
              "endIndex": 479,
              "text": "*   亚马逊推出了“前沿代理”、新型芯片以及AWS re:Invent大会上的私有“AI工厂”等产品。"
            },
            "groundingChunkIndices": [
              3
            ]
          },
          {
            "segment": {
              "startIndex": 480,
              "endIndex": 586,
              "text": "*   迪士尼向OpenAI投资10亿美元，旨在将米老鼠等迪士尼角色引入Sora AI视频工具。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 587,
              "endIndex": 690,
              "text": "*   美国前总统特朗普签署了一项行政命令，旨在限制各州对人工智能的监管。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 691,
              "endIndex": 769,
              "text": "*   亚马逊Kindle的新应用内AI功能允许用户就书籍内容提问。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 770,
              "endIndex": 920,
              "text": "*   有观点认为，虽然许多工作面临被AI颠覆的风险，但AI培训师、AI审计师和AI集成商等角色仍将保持高度相关性。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 921,
              "endIndex": 1014,
              "text": "*   OpenAI正在调整新员工的股权“归属期”条款，以留住顶尖的AI人才。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 1043,
              "endIndex": 1152,
              "text": "*   上诉法院支持对苹果公司的藐视法庭裁决，但重新开启了iPhone应用收费的讨论。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 1153,
              "endIndex": 1241,
              "text": "*   加密货币大亨Do Kwon因400亿美元的稳定币欺诈案被判处15年监禁。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 1242,
              "endIndex": 1332,
              "text": "*   福特公司正加大技术投入并与警方合作，以阻止F-150皮卡盗窃案。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 1333,
              "endIndex": 1435,
              "text": "*   Reddit对澳大利亚禁止16岁以下儿童使用社交媒体的全球首例法律提出挑战。"
            },
            "groundingChunkIndices": [
              0
            ]
          },
          {
            "segment": {
              "startIndex": 1436,
              "endIndex": 1488,
              "text": "*   OnePlus 15R手机预计将于12月17日发布。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 1489,
              "endIndex": 1578,
              "text": "*   Oppo A6x已在印度发布，配备6,500毫安时电池，起售价为12,499卢比。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 1579,
              "endIndex": 1639,
              "text": "*   iPhone 16的价格在Croma上降价超过17,000卢比。"
            },
            "groundingChunkIndices": [
              4
            ]
          },
          {
            "segment": {
              "startIndex": 1686,
              "endIndex": 1739,
              "text": "*   **QCon AI** 将于12月16-17日在纽约举行。"
            },
            "groundingChunkIndices": [
              5
            ]
          },
          {
            "segment": {
              "startIndex": 1740,
              "endIndex": 1835,
              "text": "*   **混沌通信大会 (Chaos Communication Congress)** 将于12月27-30日在汉堡举行。"
            },
            "groundingChunkIndices": [
              5
            ]
          },
          {
            "segment": {
              "startIndex": 1836,
              "endIndex": 1989,
              "text": "*   **CES 2026** 是全球最具影响力的科技盛会，将于2026年1月6日至9日在拉斯维加斯举行，展示突破性技术和全球创新。"
            },
            "groundingChunkIndices": [
              6,
              5
            ]
          },
          {
            "segment": {
              "startIndex": 1990,
              "endIndex": 2143,
              "text": "*   **GITEX GLOBAL 2025** 是全球最大的科技与人工智能盛会，第45届将于2025年10月13日在迪拜世界贸易中心和迪拜港举行。"
            },
            "groundingChunkIndices": [
              7
            ]
          },
          {
            "segment": {
              "startIndex": 2144,
              "endIndex": 2285,
              "text": "*   **新加坡科技周 (Tech Week Singapore)** 是亚洲最重要的科技盛会，将于2025年10月8日至9日在滨海湾金沙举行。"
            },
            "groundingChunkIndices": [
              7
            ]
          },
          {
            "segment": {
              "startIndex": 2286,
              "endIndex": 2353,
              "text": "*   **Web Summit** 将于2026年11月9日至12日返回里斯本。"
            },
            "groundingChunkIndices": [
              8
            ]
          },
          {
            "segment": {
              "startIndex": 2354,
              "endIndex": 2496,
              "text": "*   **LEAP 2026** 科技大会将于2026年4月13日至16日举行，同期还将举办DeepFest，这是该地区最具影响力的AI活动。"
            },
            "groundingChunkIndices": [
              9
            ]
          }
        ],
        "webSearchQueries": [
          "today's technology news",
          "major technology events recently"
        ]
      }
    }
  ],
  "usageMetadata": {
    "promptTokenCount": 13,
    "candidatesTokenCount": 723,
    "totalTokenCount": 1274,
    "promptTokensDetails": [
      {
        "modality": "TEXT",
        "tokenCount": 13
      }
    ],
    "toolUsePromptTokenCount": 99,
    "toolUsePromptTokensDetails": [
      {
        "modality": "TEXT",
        "tokenCount": 99
      }
    ],
    "thoughtsTokenCount": 439
  },
  "modelVersion": "gemini-2.5-flash",
  "responseId": "YApAab6kD_ff2roPoLjmuAU"
}
```

cto-new[bot] committed:
feat(gemini-tools): add support for google_search and code_execution tools in AxonHub transformer

This change enables Gemini googleSearch and codeExecution tools to be passed through AxonHub's Gemini transformer, ensuring real-time search and code execution capabilities can be used during generation.

Details:
- Extend the unified Tool model to include GoogleSearch and CodeExecution.
- Update inbound_convert.go to recognize googleSearch and codeExecution tools.
- Update outbound_convert.go to reconstruct google_search and code_execution tools in Gemini payloads.
- Add tests covering new tool types and mixed tool scenarios.
- Ensure existing function-tool behavior remains unaffected.

BREAKING CHANGE: none

Silentely committed:
feat(gemini): add urlContext tool support for Gemini transformer

Add support for Gemini's urlContext grounding tool, completing the full
set of Gemini grounding tools alongside googleSearch and codeExecution.

Changes:
- Add UrlContext type and field to unified LLM Tool struct
- Add UrlContext field to Gemini Tool model
- Add ToolTypeUrlContext constant
- Implement bidirectional conversion (LLM ↔ Gemini format)
- Add comprehensive unit tests for urlContext conversion

Tested:
- Unit tests: All passed
- Integration test: Successfully fetched and summarized web content
  using urlContext tool with gemini-2.5-flash model

Silentely committed:
test: 改进并发测试的同步机制

- 将并发测试中的 WaitGroup 操作重构为更清晰的语法，在每个 goroutine 前显式调用 wg.Add(1)
- 为每个 goroutine 添加 defer wg.Done() 以确保在函数返回时正确通知 WaitGroup
- 使用标准的 go func() 语法替代 wg.Go() 方法，使代码风格更一致且更易理解
- 保持测试逻辑不变，但提高了代码的可读性和同步机制的可靠性

Silentely commented:
## 流式测试结果补充

在非流式测试基础上，已完成所有三个 grounding tools 的流式测试验证：

### 测试环境
- 端点：`/gemini/v1beta/models/gemini-2.5-flash:streamGenerateContent?alt=sse`
- 响应格式：SSE (Server-Sent Events)

### 测试结果

| 工具 | 状态 | 数据块数 | 测试内容 |
|------|------|----------|----------|
| **urlContext** | ✅ 通过 | 3 块 | 解析 GitHub 网页并流式返回摘要 |
| **googleSearch** | ✅ 通过 | 12 块 | 搜索科技新闻，内容完整 |
| **codeExecution** | ✅ 通过 | 8 块 | 计算斐波那契数列前20项 |

### 响应格式验证
- ✅ 每个 chunk 以 `data: {...}` 格式返回
- ✅ 最后一个 chunk 包含 `finishReason: STOP`
- ✅ 包含完整的 `usageMetadata` 统计信息

### 示例响应 (urlContext)
```
data: {"candidates":[{"content":{"parts":[{"text":"这个 GitHub 仓库由 Anthropic 发布..."}],"role":"model"},"index":0}],...}
data: {"candidates":[{"content":{"parts":[{"text":"...评估结果。"}],"role":"model"},"index":0}],...}
data: {"candidates":[{"content":{"parts":[{"text":"简而言之..."}],"role":"model"},"finishReason":"STOP","index":0}],...}
```

**结论：流式和非流式模式下，三个 Gemini grounding tools 均正常工作。**

looplj commented:
这是啥情况，用的什么工具全部 format 了吗，golangci-lint 吗

looplj commented:
帮忙看下 CodeExecution 这一类工具应该是很多渠道都有的，参数是不是一致的，不一致的话，要不就隔离开，都增加前缀。
或者单独放到

Google  {
  Search
  CodeExecution
  XXX 
}
不放在顶层。

Silentely committed:
refactor(gemini): use GoogleTools namespace and revert code style changes

Changes:
1. Revert unnecessary code style modernizations (interface{} → any, etc.)
   to keep PR focused on feature changes only
2. Implement GoogleTools namespace for Gemini grounding tools:
   - GoogleSearch, GoogleCodeExecution, GoogleUrlContext
   - Isolates Google-specific tools from other providers
   - Improves extensibility for future provider-specific tools

Testing:
- All Gemini tools unit tests pass
- Streaming and non-streaming integration verified

looplj commented:
type 我觉得可以是 google_search, google_code_execution 这样子。
但是具体配置放到 Google 里面。

全局格式化，可以单独提一个 PR，然后可以看下是不是需要更新 lint 配置，我这怎么一直没有格式化替换 any。

Silentely commented:
已做修改，openai与Anthropic工具调用与gemini情况不同，不支持原生搜索，只支持Function Calling，但是目前有个潜在问题，工具调用仅支持原生渠道，例如gemini与vertex AI，其他通过/gemini路径调用也可能路由到OpenAI渠道上，会静默丢弃工具

looplj commented:
可以看下我上面一条 comment。


OpenAI 和 Anthropic 的原生渠道是支持 web search 工具的，不过各种第三方中转站不知道是否支持。


<img width="767" height="322" alt="image" src="https://github.com/user-attachments/assets/2a83f79c-c179-44f6-9093-69b5c8000712" />

不过，现在，可以先支持这种方式再说，应该能满足部分需求。

未来的计划
1. 转换各渠道原生的 web search 
2. AxonHub 本身集成 web search 工具，

Silentely committed:
refactor(gemini): use specific tool types instead of generic google type

Per reviewer feedback, change Tool.Type from generic "google" to specific types:
- google_search
- google_code_execution
- google_url_context

Configuration remains in Tool.Google namespace for isolation and extensibility.

Silentely committed:
fix(lint): fix godoc comment format for tool type constants

Silentely commented:
> type 我觉得可以是 google_search, google_code_execution 这样子。 但是具体配置放到 Google 里面。
> 
> 全局格式化，可以单独提一个 PR，然后可以看下是不是需要更新 lint 配置，我这怎么一直没有格式化替换 any。

tpye已做修改，全局格式化后面看看，OpenAI 和 Anthropic 的原生渠道的工具调用可以后面pr，我自己也用的少，主要是用gemini的原生搜索，new-api和one-hub的兼容应该可以，我是从one-hub转到axonhub的
